### DISCUSSION TITLE
Confused on how Cache & HTTP interact

### DISCUSSION THREAD

**Discussion by thetayloredman** [serenity-rs/serenity#2365] [Q&A] [answered]
The following code fails with "Guild not found in the cache"...

```rs
    if command == "config" {
        if !msg
            .member(&ctx.http)
            .await?
            .permissions(&ctx.cache)?
            .manage_guild()
        {
            msg.reply(
                &ctx.http,
                "You need the Manage Server permission to run this command!",
            )
            .await?;
            return Ok(());
        }
```

How do I do this such that the member is fetched only if they're not cached? Similar to some sort of Discord.js code `(await message.member.fetch()).permissions`

**Comment by arqunis** [comment, marked as answer]
[`Message::member`](https://serenity-rs.github.io/serenity/current/serenity/model/channel/struct.Message.html#method.member) accepts a parameter that implements `CacheHttp`, or in other words, a type that can provide access to the http client and/or cache. This trait is used wherever a function needs the http client, but may optionally use the cache first if it is given one. Right now, by passing `&ctx.http`, you're providing `Message::member` only the http client. You want to pass a reference to the `Context` instead: `msg.member(ctx)` (assuming `ctx` is `&Context`), as the `Context` stores both the http client and cache. This way, `Message::member` will try searching the cache for the message author's member data first before attempting a http request.

...that's not your problem, though. Your error originates from [`Member::permissions`](https://serenity-rs.github.io/serenity/current/serenity/model/guild/struct.Member.html#method.permissions). That function needs the `Guild` data to calculate the member's permissions, but it's currently failing as the `Guild` the member is in hasn't been cached yet. What gateway intents have you enabled? Do you at least have [`GatewayIntents::GUILDS`](https://serenity-rs.github.io/serenity/current/serenity/model/gateway/struct.GatewayIntents.html#associatedconstant.GUILDS)?

**Reply by thetayloredman** [reply]
I was unaware that `ctx` was supposed to be used like that. And it sure was the intents at fault. Sorry for the stupid question!
